rust/src/traits/samplers/bernoulli/mod.rs

```diff
@@ -56,37 +59,83 @@ pub trait SampleBernoulli<T>: Sized {
-        if prob == T::one() { return Ok(true) }
+        // Consider the binary expansion of prob into an infinite sequence b_i
+        //    prob = sum_{i=0}^\inf b_i / 2^(i + 1)
+        // This algorithm samples i ~ Geometric(p=0.5), then returns b_i.
+
+        // Step 1. sample first_heads_index = i ~ Geometric(p=0.5)
+        let first_heads_index = {
+            // Since prob has finite precision, there is some j for which b_i = 0 for all i > j.
+            // Thus, it is equivalent to sample i from the truncated geometric, and return false if i > j.
+            // j is the index of the last element of the binary expansion that could possibly be 1.
+            //    j = max_coin_flips
+            //      = max_{prob} [leading_zeros(prob) + mantissa_digits]
+            //      = max_{prob} [max_prob_exponent - exponent(prob) + mantissa_digits]
+            //      = max_{prob} [max_raw_prob_exponent - raw_exponent(prob) + mantissa_digits]
+            //      = max_raw_exponent + mantissa_digits
+            //               where max_raw_prob_exponent = T::EXPONENT_BIAS - 1 because prob < 1.
+            //      = (T::EXPONENT_BIAS - 1) + (T::MANTISSA_BITS + 1)
+            //      = T::EXPONENT_BIAS + T::MANTISSA_BITS
+            let max_coin_flips =
+                usize::exact_int_cast(T::EXPONENT_BIAS)? + usize::exact_int_cast(T::MANTISSA_BITS)?;
+
+            // We need to sample at least j bits. The smallest sample size is a byte. Round up to the nearest byte:
+            //    buffer_len = j.div_ceil(8)
+            // When T = f64, we sample 135 bytes.
+            //        = f32, we sample 19 bytes.
+            // If the first heads is found after j flips, but before buffer_len * 8 flips,
+            //    then it will always index into the trailing zeros of the binary expansion.
+            let buffer_len = max_coin_flips.inf_div(&8)?;
+
+            // repeatedly flip a fair coin (up to j times) to identify 0-based index i of first heads
+            match sample_geometric_buffer(buffer_len, constant_time)? {
+                // i is in terms of T::Bits, not usize; assign to first_heads_index
+                Some(i) => T::Bits::exact_int_cast(i)?,
+                // otherwise return early because i > j
+                // i is beyond the greatest possible nonzero b_i
+                None => return Ok(false),
+            }
+        };
+
+        // Step 2. index into the binary expansion of prob at first_heads_index to get b_i
 
         // number of leading zeros in binary representation of prob
-        //    cast is non-saturating because exponent only uses first 11 bits
-        //    exponent is bounded in [0, EXPONENT_PROB] by check for valid probability and one check
-        let num_leading_zeros = T::EXPONENT_PROB - prob.exponent();
+        //    exponent is bounded in [0, EXPONENT_BIAS - 1] by check for valid probability and one check
```

Comment:
The language in this comment is a bit tough to parse. Maybe say "exponent is bounded in [0, EXPONENT_BIAS - 1] by check for valid probability, and by returning when prob == 1"